src/activatable/activatable.ts

```diff
@@ -0,0 +1,130 @@
+}
+
+const LifeCycleDescription = {
+  [LifeCycleState.Created]: 'has never activated before',
+  [LifeCycleState.Activating]: 'is activating',
+  [LifeCycleState.Activated]: 'is activated',
+  [LifeCycleState.Deactivating]: 'is deactivating',
+  [LifeCycleState.Deactivated]: 'is deactivated',
+  [LifeCycleState.Disposed]: 'is disposed',
+  [LifeCycleState.FailedToActivate]: 'errored while activating',
+  [LifeCycleState.FailedToDeactivate]: 'errored while deactivating',
+};
+
+export interface IActivatable {
+  isActivating: boolean;
+  isActivated: boolean;
+  isDeactivating: boolean;
+  isDeactivated: boolean;
+  isErrored: boolean;
+  activate: () => Promise<void>;
+  deactivate: () => Promise<void>;
+  disposeActivateable: () => void;
+}
+
+export abstract class Activatable implements IActivatable {
+  private _state: LifeCycleState = LifeCycleState.Created;
+
+  public get isActivating() {
+    return this._state === LifeCycleState.Activating;
+  }
+
+  public get isActivated() {
+    return this._state === LifeCycleState.Activated;
+  }
+
+  public get isDeactivating() {
+    return this._state === LifeCycleState.Deactivating;
+  }
+
+  public get isDeactivated() {
+    return (
+      this._state === LifeCycleState.Created ||
+      this._state === LifeCycleState.Deactivated ||
+      this._state === LifeCycleState.Disposed
+    );
+  }
+
+  public get isErrored() {
+    return this._state === LifeCycleState.FailedToActivate || this._state === LifeCycleState.FailedToDeactivate;
+  }
+
+  public async activate() {
+    switch (this._state) {
+      case LifeCycleState.Activated:
+        return;
+      case LifeCycleState.Created:
+      case LifeCycleState.Deactivated:
+        try {
+          this._state = LifeCycleState.Activating;
+          await this._activate();
+          this._state = LifeCycleState.Activated;
+          return;
+        } catch (e) {
+          this._state = LifeCycleState.FailedToActivate;
+          throw e;
+        }
+      default:
+        throw new Error(`Cannot activate an Activatable that ${LifeCycleDescription[this._state]}`);
+    }
+  }
+
+  public async deactivate() {
+    switch (this._state) {
+      case LifeCycleState.Deactivated:
+        return;
+      case LifeCycleState.Activated:
+        try {
+          this._state = LifeCycleState.Deactivating;
+          await this._deactivate();
+          this._state = LifeCycleState.Deactivated;
+          return;
+        } catch (e) {
+          this._state = LifeCycleState.FailedToDeactivate;
+          throw e;
+        }
+      default:
+        throw new Error(`Cannot deactivate an Activatable that ${LifeCycleDescription[this._state]}`);
+    }
+  }
+
+  public disposeActivateable() {
+    this._state = LifeCycleState.Disposed;
+  }
+
+  protected abstract _activate(): void | Promise<void>;
+
+  protected abstract _deactivate(): void | Promise<void>;
+}
+
+export class ActivatableCollection extends Activatable implements IActivatable {
```

Comment:
Let's have a separate interface such as `IActivatableCollection`

Reply:
🤔 .... no.

That would encourage alternative implementations. Which means that the standard implementation isn't good enough. If you can think of an example why an interface makes sense, I'm all ears though?

Reply:
I'm just saying `ActivableCollection` should not implement `IActivatable`.
`IActivatableCollection` could have the same interface as `IActivatable` does for now.

Reply:
You mean we should use `extends` not `implements`?

Reply:
Well, I'd just want to have it less tied to `IActivatable`, but lets leave it as is..

Reply:
Hmm. Well the whole point is that an `ActivatableCollection` is an `IActivatable` itself. That's what enables hierarchies.